crates/wasi-common/src/sys/unix/bsd/osfile.rs

```diff
@@ -0,0 +1,106 @@
+use crate::sys::oshandle::AsFile;
+use crate::wasi::Result;
+use std::cell::{Cell, RefCell, RefMut};
+use std::fs::File;
+use std::io;
+use std::mem::ManuallyDrop;
+use std::os::unix::prelude::{AsRawFd, FromRawFd, IntoRawFd, RawFd};
+use yanix::dir::Dir;
+
+#[derive(Debug)]
+pub(crate) struct OsFile {
+    fd: Cell<RawFd>,
+    // In case that this `OsHandle` actually refers to a directory,
+    // when the client makes a `fd_readdir` syscall on this descriptor,
+    // we will need to cache the `libc::DIR` pointer manually in order
+    // to be able to seek on it later. While on Linux, this is handled
+    // by the OS, BSD Unixes require the client to do this caching.
+    //
+    // This comes directly from the BSD man pages on `readdir`:
+    //   > Values returned by telldir() are good only for the lifetime
+    //   > of the DIR pointer, dirp, from which they are derived.
+    //   > If the directory is closed and then reopened, prior values
+    //   > returned by telldir() will no longer be valid.
+    dir: RefCell<Option<Dir>>,
+}
+
+impl OsFile {
+    /// Consumes `other` taking the ownership of the underlying
+    /// `RawFd` file descriptor.
+    pub(crate) fn update_from(&self, other: Self) {
```

Comment:
That's an excellent observation, thanks! I think adding a comment to remember about this potential issue is the best thing to do here.

On a related albeit somewhat different note, I'd still like to refactor how we store the `DIR` stream pointer on BSDes in `OsFile`. Having an embedded `Option` seems like a less than ideal solution to use here. I was actually thinking of splitting the type into two or possibly an enum, so that if we are dealing with a dir on BSD, we'd immediately store a valid `DIR` stream pointer. But this will be a subsequent PR :-)